python/dgl/graphbolt/item_sampler.py

```diff
@@ -326,8 +329,104 @@ def __iter__(self) -> Iterator:
         data_pipe = data_pipe.collate(collate_fn=self._collate)
 
         # Map to minibatch.
-        data_pipe = data_pipe.map(
-            partial(self._minibatcher, names=self._item_set.names)
+        data_pipe = data_pipe.map(partial(self._minibatcher, names=self._names))
+
+        return iter(data_pipe)
+
+
+@functional_datapipe("distributed_sample_item")
+class DistributedItemSampler(ItemSampler):
+    """Distributed Item Sampler.
+
+    This is useful with torch's DDP training. It will create distributed item
+    subset of data which could be node IDs, node pairs with or without labels,
+    node pairs with negative sources/destinations, DGLGraphs and heterogeneous
+    counterparts. The items of the original item set will be sharded and form
+    an exclusive subset for each replica (process).
+
+    Note: This class `DistributedItemSampler` is not decorated with
+    `torchdata.datapipes.functional_datapipe` on purpose. This indicates it
+    does not support function-like call. But any iterable datapipes from
+    `torchdata` can be further appended.
+
+    Parameters
+    ----------
+    item_set : ItemSet or ItemSetDict
+        Data to be sampled.
+    batch_size : int
+        The size of each batch.
+    minibatcher : Optional[Callable]
+        A callable that takes in a list of items and returns a `MiniBatch`.
+    drop_last : bool
+        Option to drop the last batch if it's not full.
+    shuffle : bool
+        Option to shuffle before sample.
+    num_replicas: int
+        Number of processes participating in distributed training. By default,
+        `world_size` is retrieved from the current distributed group.
+    even_inputs : bool
+        Option to make sure the numbers of batches for each replica are the
+        same. This will be done by dropping the excessive batches of some
+        replicas.
+
+    Examples
+    --------
+    """
+
+    def __init__(
+        self,
+        item_set: ItemSet or ItemSetDict,
+        batch_size: int,
+        minibatcher: Optional[Callable] = minibatcher_default,
+        drop_last: Optional[bool] = False,
+        shuffle: Optional[bool] = False,
+        num_replicas: Optional[int] = None,
+        even_inputs: Optional[bool] = False,
+    ) -> None:
+        super().__init__(item_set, batch_size, minibatcher, drop_last, shuffle)
+        self._even_inputs = even_inputs
+        # Apply a sharding filter to distribute the items.
+        self._item_set = self._item_set.sharding_filter()
+        # Get world size.
+        if num_replicas is None:
+            if not dist.is_available():
+                raise RuntimeError(
+                    "Requires distributed package to be available"
+                )
+            num_replicas = dist.get_world_size()
```

Comment:
What if input num_replicas is conflict with dist.get_world_size()?

Reply:
It will use the given `num_replicas`.

Reply:
Let's me make the question clear, in which case num_replicas does not equal to get_world_size.
Intuitively we don't need to pass in num_replicas since we can get_world_size. The comment should focus on address this confusion.

Reply:
It could be used when `dist` is not available during the instantiation of `DistributedItemSampler`. Most of the cases we don't have to use this param. But the comment is clear I think: we only retrieve it from `dist.get_world_size()` when `num_replicas` is None.